### PR TITLE
[WIP] Initial changes for cloud-provider with GCE.

### LINKED ISSUE #48690: Cloud Controller Manager Beta Requirements
Forked from https://github.com/kubernetes/features/issues/88 in order to allow more discussion.

Things needed for beta
-----------------------------------

1. Lots of docs (tutorials, blogs, walkthroughs etc)
2. Testing with various clouds (dogfooding)
3. Automatically build and create a image for cloud-controller-manager(CCM) on gcr.io
4. Automatically run e2e tests for CCM 
5. bugfix for kubelet certificate ip addresses (discussed here - https://github.com/kubernetes/kubernetes/pull/47152#issuecomment-307836402)
6. Make *ByID methods for all cloud integrations: https://github.com/kubernetes/kubernetes/issues/47257
7. Make possible to use an Initializer instead of the PersistentVolumeLabel admission plugin for out-of-tree clouds?

Nice to haves:
----------------------------
1. Examples of out-of-tree cloud providers (DigitalOcean, Oracle, Rancher, etc)
2. ...

Very much help wanted.
cc @wlan0 @thockin @roberthbailey @rrati @dims @andrewsykim @munnerz

### PR DESCRIPTION
**Pull request by cheftako**
This is just the initial set of changes needed to get it up.
With this the cloud-controller-manager will start in GCE with
cluster/kube-up.sh
In addition the service seems to come up in healthy state with the
permissions it needs.
Still needs considerable testing to make sure it is essentially working.

**What this PR does / why we need it**: Work in progress for enabling GCE to be in stage1 of enabling cloud-provider

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #48690

**Special notes for your reviewer**: Work in progress.

**Release note**: ```To be done```

### CODE REVIEW

File: cluster/gce/gci/configure-helper.sh [hunk 1]
```diff
@@ -1441,6 +1469,64 @@ function start-kube-apiserver {
...
+function start-cloud-controller-manager {
+  echo "Start cloud provider controller-manager"
+  create-cloudcontrollermanager-kubeconfig
+  prepare-log-file /var/log/cloud-controller-manager.log
+  # Calculate variables and assemble the command line.
+  local params="${CONTROLLER_MANAGER_TEST_LOG_LEVEL:-"--v=2"} ${CLOUD_CONTROLLER_MANAGER_TEST_ARGS:-} ${CLOUD_CONFIG_OPT}"
+  params+=" --use-service-account-credentials"
+  params+=" --cloud-provider=gce"
+  params+=" --kubeconfig=/etc/srv/kubernetes/cloud-controller-manager/kubeconfig"
+  params+=" --service-account-private-key-file=${SERVICEACCOUNT_KEY_PATH}"
```
Comment (wlan0) [inline cluster/gce/gci/configure-helper.sh]: This parameter is no-op. This can be removed

**Comment by wlan0** [comment]
@cheftako Thanks for doing this! This is awesome :) 

In order to get the e2e tests to pass, they need to wait for the ccm to set the labels instead of the kubelet setting it. We might need some refactoring of the e2e tests.

Other than that, Just one small comment from my side.

File: cluster/gce/container-linux/configure-helper.sh [hunk 2]
```diff
@@ -1069,7 +1179,8 @@ function start-kube-controller-manager {
   # Calculate variables and assemble the command line.
   local params="${CONTROLLER_MANAGER_TEST_LOG_LEVEL:-"--v=2"} ${CONTROLLER_MANAGER_TEST_ARGS:-} ${CLOUD_CONFIG_OPT}"
   params+=" --use-service-account-credentials"
-  params+=" --cloud-provider=gce"
+  params+=" --cloud-provider=external"
+  params+=" --external-plugin=gce"
```
Comment (colemickens) [inline cluster/gce/container-linux/configure-helper.sh]: What is this used for? Is this for some short-term shim?

File: cmd/kubelet/app/server.go [hunk 3]
```diff
@@ -305,7 +305,15 @@ func run(s *options.KubeletServer, kubeDeps *kubelet.Dependencies) (err error) {
 				glog.V(2).Infof("Successfully initialized cloud provider: %q from the config file: %q\n", s.CloudProvider, s.CloudConfigFile)
 			}
 			kubeDeps.Cloud = cloud
+		} else if cloudprovider.IsExternal(s.CloudProvider) {
+			glog.V(2).Infof("WRF - Initting the cloud provider as GCE anyway!!!")
```
Comment (colemickens) [inline cmd/kubelet/app/server.go]: Is this for temp debugging? If not, what's this for?

File: cluster/gce/container-linux/configure-helper.sh (hunk 2 above)
Comment (cheftako) [reply cluster/gce/container-linux/configure-helper.sh]: Essentially yes. VolumeController is not in the current round of extraction(https://goo.gl/CtzpVm). However VolumeController requires a working cloud. Unfortunately the relevant flag is conflating which cloud to use with which controllers to run. This is a slightly clumsy separation of which cloud provider to load from which plugins to run. When I did the original wiring I did not realize how deeply we had conflated these concepts. So right now I'm trying to make sure I can successfully break them. Once that is done I will come back and work out how to clean up the flags. Eventually we should be able to drop the need to specify and cloud provider. However I believe we will always need a way to specify which controllers to run.

File: cmd/kubelet/app/server.go (hunk 3 above)
Comment (cheftako) [reply cmd/kubelet/app/server.go]: Debugging. This is a WIP. So far when I generate something which can be split off and used in isolation I am generating a separate PR for that. I may eventually attempt to "clean" this PR but that will be quite a bit of work.

**Comment by enxebre** [comment]
Hey @cheftako - "Work in progress for enabling GCE to be in stage1" is there a place where I can see what are the different stages? As part of this refactor is there any plan to solve the problem arising because of https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_network.go#L85 which is stated here https://github.com/kubernetes-incubator/bootkube/issues/463#issuecomment-337480735

**Comment by cheftako** [comment]
Hi @enxebre, sorry I missed your question. So https://github.com/kubernetes/community/blob/master/contributors/design-proposals/cloud-provider/cloud-provider-refactoring.md used to discuss the stages but that no longer seems to be true. So the TLDR version is stage 1 is all of the cloud provider (i.e. the 7 cloud providers currently in kubernetes/kubernetes) specific code should be moved and run in cloud-controller-manager. So if you don't want to run any of their code you don't run the cloud-controller-manager and you don't run their code. In stage 2 the cloud provider can then build their own version of the cloud-controller-manager which has only their plugin in it. That way they don't have to worry about running any other cloud providers code.

**Comment by luxas** [comment]
@cheftako What's the status of this PR?

**Comment by cheftako** [comment]
@luxas Its a permanent [WIP]. It was my proof of concept/scratch space for getting GCE working with the only cloud-provider code being in the CCM. I'm now trying to extract the pieces of the solution found here, clean them up and push them as working PRs. https://github.com/kubernetes/kubernetes/pull/57492 and https://github.com/kubernetes/kubernetes/pull/52371 are examples of that extraction. I should rebase this PR to help me track what remains to be done and ensure that the remaining issues are clear.

**Pull request closed without merging**